test/memory_tools/test_memory_tools.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include "./memory_tools.hpp"
+
+/* Tests the allocatation checking tools.
+ */
+TEST(TestMemoryTools, test_allocation_checking_tools) {
+  size_t unexpected_mallocs = 0;
+  auto on_unexpected_malloc = ([&unexpected_mallocs]() {
+    unexpected_mallocs++;
+  });
+  set_on_unexpected_malloc_callback(on_unexpected_malloc);
+  size_t unexpected_reallocs = 0;
+  auto on_unexpected_realloc = ([&unexpected_reallocs]() {
+    unexpected_reallocs++;
+  });
+  set_on_unexpected_realloc_callback(on_unexpected_realloc);
+  size_t unexpected_frees = 0;
+  auto on_unexpected_free = ([&unexpected_frees]() {
+    unexpected_frees++;
+  });
+  set_on_unexpected_free_callback(on_unexpected_free);
+  void * mem = nullptr;
+  void * remem = nullptr;
+  // First try before enabling, should have no effect.
+  mem = malloc(1024);
+  ASSERT_NE(mem, nullptr);
+  remem = realloc(mem, 2048);
+  ASSERT_NE(remem, nullptr);
+  if (!remem) {free(mem);}
```

Comment:
Question: redundant condition? previous assert would have raised if this condition was not true, same below